cmd/bucket-replication.go

```diff
@@ -407,83 +451,138 @@ func getReplicationAction(oi1 ObjectInfo, oi2 minio.ObjectInfo) replicationActio
 		oi1.VersionID != oi2.VersionID ||
 		oi1.Size != oi2.Size ||
 		oi1.DeleteMarker != oi2.IsDeleteMarker ||
-		!oi1.ModTime.Equal(oi2.LastModified) {
+		oi1.ModTime.Unix() != oi2.LastModified.Unix() {
 		return replicateAll
 	}
+
 	if oi1.ContentType != oi2.ContentType {
 		return replicateMetadata
 	}
+
 	if oi1.ContentEncoding != "" {
 		enc, ok := oi2.Metadata[xhttp.ContentEncoding]
-		if !ok || strings.Join(enc, "") != oi1.ContentEncoding {
+		if !ok {
+			enc, ok = oi2.Metadata[strings.ToLower(xhttp.ContentEncoding)]
+			if !ok {
+				return replicateMetadata
+			}
+		}
+		if strings.Join(enc, "") != oi1.ContentEncoding {
 			return replicateMetadata
 		}
 	}
+
+	// Do not compare oi2.UserMetadata - it does not contain `x-amz-meta-` prefix.
+
 	// compare metadata on both maps to see if meta is identical
 	for k1, v1 := range oi1.UserDefined {
-		if v2, ok := oi2.UserMetadata[k1]; ok && v1 == v2 {
+		if v2, ok := oi2.Metadata[strings.ToLower(k1)]; ok && v1 == strings.Join(v2, "") {
 			continue
 		}
 		if v2, ok := oi2.Metadata[k1]; ok && v1 == strings.Join(v2, "") {
 			continue
 		}
 		return replicateMetadata
 	}
-	for k1, v1 := range oi2.UserMetadata {
-		if v2, ok := oi1.UserDefined[k1]; !ok || v1 != v2 {
-			return replicateMetadata
-		}
-	}
+
 	for k1, v1slc := range oi2.Metadata {
 		v1 := strings.Join(v1slc, "")
-		if k1 == xhttp.ContentEncoding { //already compared
+		if strings.EqualFold(k1, xhttp.ContentEncoding) ||
```

Comment:
These make me worry a bit. There are known vulnerabilities where using folding produces false matches and potential conflicting values for the "same" metadata.

Here is an example where the Kelvin unicode matches and produces different values based on which one is processed first: https://play.golang.org/p/g4XS3dUJ5vm

I would like @aead to review this.

Reply:
For me right now there is no other way other than repeatedly doing this twice. For both cases.

Reply:
HTTP headers name don't support unicode AFAIK @klauspost so not sure if this will ever occur in the first place. But to be safe I can write a different function that matches these twice.